binding-macro/src/schema.rs

```diff
@@ -0,0 +1,227 @@
+                .expect("#[description]: comments should be string");
+            if is_field {
+                return Some(format!("  # {}\n", comment.value));
+            } else {
+                return Some(format!("# {}\n", comment.value));
+            }
+        }
+    }
+    None
+}
+
+fn generate_field_token(f: &Field, ty: &Ident, is_vec: bool) -> proc_macro2::TokenStream {
+    let f_str = f
+        .ident
+        .as_ref()
+        .expect("#[derive(SchemaObject)]: struct fields should be named")
+        .to_string();
+
+    let mut field_str = "".to_owned();
+    let comment = extract_comment(&f.attrs, true).unwrap_or_default();
+    field_str.push_str(comment.as_str());
+    let fmt_str = if is_vec {
+        "  {}: [{}!]!\n"
+    } else {
+        "  {}: {}!\n"
+    };
+    field_str.push_str(fmt_str);
+
+    quote! {
+        let ty_name = #ty::name();
+        let field_str = format!(#field_str, #f_str, ty_name);
+        schema.push_str(field_str.as_str());
+        #ty::schema(register);
+    }
+}
+
+fn extract_ident_from_path(ty: &Type) -> Ident {
+    match ty {
+        Type::Path(path) => path
+            .path
+            .segments
+            .first()
+            .expect("#[derive(SchemaObject)]: Vec should contain a type arg")
+            .ident
+            .clone(),
+        _ => panic!("#[derive(SchemaObject)]: Vec arg should be a path type"),
+    }
+}
+
+fn extract_ident_from_ty(ty: &Type) -> (Ident, bool) {
+    match ty {
+        Type::Path(ty) => {
+            let segs = &ty.path.segments;
+            if 1 == segs.len() {
+                let concrete_ty = segs.first().unwrap();
+                if "Vec" == &concrete_ty.ident.clone().to_string() {
+                    if let PathArguments::AngleBracketed(g_ty) = &concrete_ty.arguments {
+                        let arg = g_ty
+                            .args
+                            .first()
+                            .expect("#[derive(SchemaObject)]: Vec should contain a type arg");
+                        if let GenericArgument::Type(arg_ty) = arg {
+                            let ident = extract_ident_from_path(&arg_ty);
+                            (ident, true)
+                        } else {
+                            panic!("#[derive(SchemaObject)]: Vec arg should be a type")
+                        }
+                    } else {
+                        panic!("#[derive(SchemaObject)]: Vec should be AngleBracketed")
+                    }
+                } else if let PathArguments::None = concrete_ty.arguments {
+                    (concrete_ty.ident.clone(), false)
+                } else {
+                    panic!("#[derive(SchemaObject)]: field type only supports T, Vec<T>")
+                }
+            } else {
+                panic!("#[derive(SchemaObject)]: length of field type should be 1")
+            }
+        }
+        _ => panic!("#[derive(SchemaObject)]: field type only supports T, Vec<T>"),
+    }
+}
```

Comment:
It is hard to read for the deep indent. Try to make it flatter.

Reply:
In this circumstance, i think it's hard to do that. Can you provide some specifics ?

Reply:
1. Move the `else` part forward.

```
// change
if xx {
  // long code
} else {
  panic!
}

// to
if !xx {
  panic
}
// long code
```

2. Extract the inner variable first and then write the long logic code, other than writing it inside the block. 

```
// change
match ty {
    Type::Path(ty) => {
       // long code
    },
    _ => panic!("#[derive(SchemaObject)]: field type only supports T, Vec<T>"),
};

// to
let inner_ty = match ty {
    Type::Path(ty) => ty,
    _ => panic!("#[derive(SchemaObject)]: field type only supports T, Vec<T>"),
};
// long code using inner_ty
```

Reply:
@huwenchao  You still owe me a task, the muta-bot's weekly report isn't finished yet  🤪

Reply:
Nice move, I will change to that